packages/core/src/engine/Autodiff.ts

```diff
@@ -1747,6 +1752,149 @@ const genBytes = (graphs: ad.Graph[]): Uint8Array => {
   return mod.bytes;
 };
 
+interface Metadata {
+  numInputs: number;
+  numSecondary: number;
+
+  offsetInputs: number;
+  offsetMask: number;
+  offsetGradient: number;
+  offsetSecondary: number;
+  offsetStack: number;
+
+  memory: WebAssembly.Memory;
+
+  arrInputs: Float64Array;
+  arrMask: Int32Array;
+  arrGrad: Float64Array;
+  arrSecondary: Float64Array;
+}
+
+const makeMeta = (graphs: ad.Graph[]): Metadata => {
+  const offsetInputs = 0;
+  const numInputs = Math.max(
+    0,
+    ...graphs.flatMap(({ graph }) =>
+      getInputs(graph).map(({ label: { key } }) => key + 1)
+    )
+  );
+
+  const offsetMask = offsetInputs + numInputs * bytesF64;
+
+  const offsetGradient = offsetMask + Math.ceil(graphs.length / 2) * bytesF64;
+
+  const offsetSecondary = offsetGradient + numInputs * bytesF64;
+  const numSecondary = Math.max(0, ...graphs.map((g) => g.secondary.length));
+
+  const offsetStack = offsetSecondary + numSecondary * bytesF64;
+
+  // each WebAssembly memory page is 64 KiB, and we add one more for the stack
+  const memory = new WebAssembly.Memory({
+    initial: Math.ceil(offsetStack / (64 * 1024)) + 1,
+  });
+  const { buffer } = memory;
+
+  return {
+    numInputs,
+    numSecondary,
+
+    offsetInputs,
+    offsetMask,
+    offsetGradient,
+    offsetSecondary,
+    offsetStack,
+
+    memory,
+
+    arrInputs: new Float64Array(buffer, offsetInputs, numInputs),
+    arrMask: new Int32Array(buffer, offsetMask, graphs.length),
+    arrGrad: new Float64Array(buffer, offsetGradient, numInputs),
+    arrSecondary: new Float64Array(buffer, offsetSecondary, numSecondary),
+  };
+};
+
+const makeImports = (memory: WebAssembly.Memory): WebAssembly.Imports => ({
+  [importModule]: {
+    [importMemoryName]: memory,
+    ...Object.fromEntries(
+      [...builtins.keys()].map((name, i) => [
+        i.toString(36),
+        {
+          inverse: (x: number): number => 1 / x,
+
+          acos: Math.acos,
+          acosh: Math.acosh,
+          asin: Math.asin,
+          asinh: Math.asinh,
+          atan: Math.atan,
+          atanh: Math.atanh,
+          cbrt: Math.cbrt,
+          cos: Math.cos,
+          cosh: Math.cosh,
+          exp: Math.exp,
+          expm1: Math.expm1,
+          log: Math.log,
+          log1p: Math.log1p,
+          log10: Math.log10,
+          log2: Math.log2,
+          sign: Math.sign,
+          sin: Math.sin,
+          sinh: Math.sinh,
+          tan: Math.tan,
+          tanh: Math.tanh,
+
+          atan2: Math.atan2,
+          pow: Math.pow,
+
+          polyRoots: (p: number, n: number): void => {
+            polyRoots(new Float64Array(memory.buffer, p, n));
+          },
+        }[name],
+      ])
+    ),
+  },
+});
+
+const makeCompiled = (
+  graphs: ad.Graph[],
+  meta: Metadata,
+  instance: WebAssembly.Instance
+): Gradient => {
+  const f = instance.exports[exportFunctionName] as (
```

Comment:
Can you document why `f` here takes a bunch of `number`s whereas `GradFn` (i.e. `wrapped` below) takes arrays of floats or ints? I'm assuming this has something to do with how `WebAssembly.Instance.exports` works? The [docs on exports](https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/Instance/exports) are kinda sparse for me to understand.

Reply:
I wrote comments explaining these two signatures; does this help?

Reply:
Gotcha. I wasn't sure they were supposed to be pointers because the variable names said `offset` a lot. Makes sense to me!

Reply:
Yeah I guess the words `offset` and `pointer` are fairly related but not directly synonymous 🤔